doc/weakly_supervised.rst

```diff
@@ -0,0 +1,316 @@
+3D array-like of tuples of shape ``(n_tuples, t, n_features)``, where
+``n_tuples`` is the number of tuples, ``tuple_size`` is the number of elements
+in a tuple (2 for pairs, 3 for triplets for instance), and ``n_features`` is
+the number of features of each point.
+
+.. rubric:: Example:
+   Here is an artificial dataset of 4 pairs of 2 points of 3 features each:
+
+>>> import numpy as np
+>>> tuples = np.array([[[-0.12, -1.21, -0.20],
+>>>                     [+0.05, -0.19, -0.05]],
+>>>
+>>>                    [[-2.16, +0.11, -0.02],
+>>>                     [+1.58, +0.16, +0.93]],
+>>>
+>>>                    [[+1.58, +0.16, +0.93 ],  # same as tuples[1, 1, :]
+>>>                     [+0.89, -0.34, +2.41]],
+>>>
+>>>                    [[-0.12, -1.21, -0.20 ],  # same as tuples[0, 0, :]
+>>>                     [-2.16, +0.11, -0.02]]])  # same as tuples[1, 0, :]
+>>> y = np.array([-1, 1, 1, -1])
+
+.. warning:: This way of specifying pairs is not recommended for a large number
+   of tuples, as it is redundant (see the comments in the example) and hence
+   takes a lot of memory. Indeed each feature vector of a point will be
+   replicated as many times as a point is involved in a tuple. The second way
+   to specify pairs is more efficient
+
+
+2D array of indicators + preprocessor
+-------------------------------------
+
+Instead of forming each point in each tuple, a more efficient representation
+would be to keep the dataset of points ``X`` aside, and just represent tuples
+as a collection of tuples of *indices* from the points in ``X``. Since we loose
+the feature dimension there, the resulting array is 2D.
+
+.. rubric:: Example: An equivalent representation of the above pairs would be:
+
+>>> X = np.array([[-0.12, -1.21, -0.20],
+>>>               [+0.05, -0.19, -0.05],
+>>>               [-2.16, +0.11, -0.02],
+>>>               [+1.58, +0.16, +0.93],
+>>>               [+0.89, -0.34, +2.41]])
+>>>
+>>> tuples_indices = np.array([[0, 1],
+>>>                            [2, 3],
+>>>                            [3, 4],
+>>>                            [0, 2]])
+>>> y = np.array([-1, 1, 1, -1])
+
+In order to fit metric learning algorithms with this type of input, we need
+to give the original dataset of points ``X`` to the estimator so that it
+knows what point the indices refer to. We do this when initializing the
+estimator, through the argument `preprocessor`.
+
+.. rubric:: Example:
+
+>>> from metric_learn import MMC
+>>> mmc = MMC(preprocessor=X)
+>>> mmc.fit(pairs_indice, y)
+
+
+.. note::
+
+   Instead of an array-like, you can give a callable in the argument
+   ``preprocessor``, which will go fetch and form the tuples. This allows to
+   give more general indicators than just indices from an array (for instance
+   paths in the filesystem, name of records in a database etc...) See section
+   :ref:`preprocessor` for more details on how to use the preprocessor.
```

Comment:
link to section does not seem to work (at least not on the version I compiled locally)